### PR TITLE
fix(assets): parallel docker image publishing fails on macOS

### LINKED ISSUE #20116: (cdk-assets): osx cannot run identical `docker login` commands in parallel
### Describe the bug

Per @brianfrantz in https://github.com/aws/aws-cdk/pull/19367#issuecomment-1102691605:

> It appears that this change has exposed a race condition when publishing multiple ECS images from a Mac. It is causing our deployments to fail with:
> 
> `[83%] fail: docker login --username AWS --password-stdin https://285872988623.dkr.ecr.us-east-1.amazonaws.com exited with error code 1: Error saving credentials: error storing credentials - err: exit status 1, out: The specified item already exists in the keychain.`
> 
> It appears that this issue is caused by a combination of the CDK performing a “docker login” on each published asset, along with specific logic in the docker project for storing credentials in the OSX keychain.
> 
> The source code for [docker-credentials-osxkeychain](https://github.com/docker/docker-credential-helpers/blob/master/osxkeychain/osxkeychain_darwin.go#L35) doesn’t update existing OSX keychain entries if present. Instead it always does a [SecKeychainAddInternetPassword](https://developer.apple.com/documentation/security/1393322-seckeychainaddinternetpassword) call, which fails if the entry is already in the keychain. That error isn’t caught and handled. Instead the code attempts to avoid this issue by deleting the entry in the keychain before each add attempt. This logic works fine for the normal use case of doing a single login and then making multiple authenticated docker calls (single-threaded or in parallel). In fact, the logic has been in place for five-plus years.
> 
> Unfortunately, the [CDK calls “docker login” before each “docker push” call](https://github.com/aws/aws-cdk/blob/master/packages/cdk-assets/lib/private/handlers/container-images.ts#L19) instead of once for the batch. Two publish calls started at the same time will (most of the time) result in a credentials sequence of “delete, delete, add, add” such that the second add will fail because the first add created the entry.
> 
> The best method for fixing this issue in the CDK would involve refactoring the publishing code to perform a single “docker login” for the batch of container assets. Unfortunately this would be a significant redesign as the logic is shared between lambda assets and container assets.
> 
> Another option might be to catch the failed login and try again after a small delay, with or without detecting the specific error message. This doesn’t feel super clean, but is less effort and is less likely to cause side effects. That said, while it should work fine for small numbers of container assets, it may become problematic when there are many publish calls all logging in and retrying all at once.

### Expected Behavior

`docker login` should be run once per ecr repository any docker image asset is being pushed to.

### Current Behavior

It runs `docker login` for every push, concurrently, and this causes the login command to fail.

### Reproduction Steps

**app.ts**
```ts
import { App, Stack } from 'aws-cdk-lib';
import { ContainerImage, FargateTaskDefinition } from 'aws-cdk-lib/aws-ecs';

const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

const stack = new Stack(app, 'cdk-concurrency-bug', { env: devEnv });

for (let i = 0; i < 30; i++) {
  const task = new FargateTaskDefinition(stack, `Task${i}`, {});
  task.addContainer('asset', {
    image: ContainerImage.fromAsset(__dirname, {
      buildArgs: {
        VAR: `var-${i}`,
      },
    }),
  });
}

app.synth();
```

**Dockerfile**
```
FROM debian:11-slim
ARG VAR=foo

WORKDIR /app
RUN echo $VAR >var.txt
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### CDK CLI Version

2.21.1

### Framework Version

2.21.1

### Node.js Version

14.19.1

### OS

OSX

### Language

Typescript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by misterjoshua**
Changes container image publishing so that publishing doesn't re-run docker logins for the same repository and so logins are run one at a time.

Fixes #20116

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by mitchlloyd** [comment]
I'd like to help by testing against my app since I can consistently get a failure with the current version of CDK. However, I'm having a heck of a time trying to test against the CDK source. I ran

```
scripts/foreach.sh yarn install --frozen-lockfile
scripts/foreach.sh yarn build
# from my project root
../../aws/aws-cdk/link-all.sh
```

But then I'm getting stuck with the fact that the built `aws-cdk-lib` requires a peerDependency on constructs 3.x while my app (and current CDK version) is using 10.x. Attempting to downgrade to 3.x in my app doesn't seem to be working as `Construct` has no public methods. Attempting to upgrade to 10.x in the cdk libraries doesn't seem to work because I end up with type conflicts between `<my-project>/node_modules/aws-cdk-lib/core/lib/construct-compat).IConstruct` and `<my-project>/node_modules/constructs/lib/construct").IConstruct`.

Any idea how to work around this 3.x/10.x mismatch?

**Comment by misterjoshua** [comment]
> I'd like to help by testing against my app since I can consistently get a failure with the current version of CDK. However, I'm having a heck of a time trying to test against the CDK source. I ran
> 
> ```
> scripts/foreach.sh yarn install --frozen-lockfile
> scripts/foreach.sh yarn build
> # from my project root
> ../../aws/aws-cdk/link-all.sh
> ```
> 
> But then I'm getting stuck with the fact that the built `aws-cdk-lib` requires a peerDependency on constructs 3.x while my app (and current CDK version) is using 10.x. Attempting to downgrade to 3.x in my app doesn't seem to be working as `Construct` has no public methods. Attempting to upgrade to 10.x in the cdk libraries doesn't seem to work because I end up with type conflicts between `<my-project>/node_modules/aws-cdk-lib/core/lib/construct-compat).IConstruct` and `<my-project>/node_modules/constructs/lib/construct").IConstruct`.
> 
> Any idea how to work around this 3.x/10.x mismatch?

I usually go like this to globally link `cdk`:

```
yarn --immutable
cd packages/aws-cdk
../../scripts/buildup
yarn link
```

Then in my project I just type `yarn link aws-cdk`. In this case, you can go `yarn cdk --version` and it should show version `0.0.0`.

**Comment by mitchlloyd** [comment]
Thanks for the detail instructions @misterjoshua. Relying on `yarn link` worked a lot better than what I was digging up from the internal CDK documentation.

I can confirm that this fixes the issue I saw in my application when I switch to the `aws-cdk` library from this PR.

---

I updated to 2.21.1 and added a dummy `COPY` command in 2 of my Docker files to bust the cache and get all new ECS task definitions.

I got this error:

```
[66%] fail: docker login --username <redacted> --password-stdin <redacted> exited with error code 1: Error saving credentials: error storing credentials - err: exit status 1, out: `error storing credentials - err: exit status 1, out: `The specified item already exists in the keychain.``
```

Checkout `fix-multi-login` branch and run

```
yarn --immutable
cd packages/aws-cdk
../../scripts/buildup
yarn link
```

In my project

```
yarn link aws-cdk
yarn cdk --version # => 0.0.0 (build f3fd3d3)
```

Then my CDK deployment worked as expected.

**Comment by misterjoshua** [comment]
@pahud Hey. This PR could use a little review. Are you still running that community reviewer program?

**Comment by misterjoshua** [comment]
@pahud Hey. This PR could use a little review. Are you still running that community reviewer program?

**Comment by brianfrantz** [comment]
@misterjoshua @mitchlloyd Sorry.  Only today got to a place where I could test this.  Running cdk from this branch worked as expected.  Running cdk version 2.23.0 (build 50444aa) still returns "The specified item already exists in the keychain."

Best I can tell, this fix will solve our issue.

File: packages/cdk-assets/lib/private/util.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Creates a critical section, ensuring that at most one function can
+ * enter the critical section at a time.
+ */
+export function createCriticalSection() {
+  let lock = Promise.resolve();
+  return async (criticalFunction: () => Promise<void>) => {
+    const res = lock.then(() => criticalFunction());
+    lock = res.catch(e => e);
+    return res.then(v => v);
```
Comment (rix0rrr) [inline packages/cdk-assets/lib/private/util.ts:10]: Couldn't this be `return res` ? What's the value of the added `then` ?

Comment (misterjoshua) [reply packages/cdk-assets/lib/private/util.ts]: Good spotting this. I've removed it - it was a leftover from some hand testing I was doing.

**Pull request merged**